clippy_lints/src/matches/mod.rs

```diff
@@ -1118,8 +1119,13 @@ impl<'tcx> LateLintPass<'tcx> for Matches {
                 );
                 needless_match::check_if_let(cx, expr, &if_let);
             }
-        } else if !from_expansion {
-            redundant_pattern_match::check(cx, expr);
+        } else {
+            if let Some(while_let) = higher::WhileLet::hir(expr) {
+                significant_drop_in_scrutinee::check_while_let(cx, expr, while_let.let_expr, while_let.if_then);
+            }
+            if !from_expansion {
+                redundant_pattern_match::check(cx, expr);
+            }
         }
```

Comment:
nit (feel free to ignore). Could this be written as:

```rust
} else if let Some(while_let) = higher::WhileLet::hir(expr) {
    if from_expansion {
        significant_drop_in_scrutinee::check_while_let(cx, expr, while_let.let_expr, while_let.if_then);
    } else {
        redundant_pattern_match::check(cx, while_let);
    }
}
```
then remove the other `higher::WhileLet::hir(expr)` within `redundant_pattern_match::check`, it improves readability for a little bit, plus, it's literally a "redundant pattern matching" 😂